internal/pkg/skuba/upgrade/addon/version.go

```diff
@@ -79,9 +81,15 @@ func HasAddonUpdate(aviu AddonVersionInfoUpdate) bool {
 }
 
 func hasAddonManifestUpdateWithAddon(aviu AddonVersionInfoUpdate, addon kubernetes.Addon) bool {
+	if aviu.Current[addon] == nil {
+		return true
+	}
 	return aviu.Updated[addon].ManifestVersion > aviu.Current[addon].ManifestVersion
 }
 
 func hasAddonVersionUpdateWithAddon(aviu AddonVersionInfoUpdate, addon kubernetes.Addon) bool {
+	if aviu.Current[addon] == nil {
+		return true
+	}
 	return aviu.Updated[addon].Version > aviu.Current[addon].Version
```

Comment:
I seem to have missed this one on a previous review (!!), this comparison is not right. We are depending on string comparisons here (https://play.golang.org/p/mP5fBhm2_kb), but we should parse the semantic version to a `*version.Version`, and use `Compare` or `LessThan`. We can follow up on this one on another PR if you prefer.

Reply:
Already found this issue and I prefer file another PR to address it.

Reply:
This looks like a blocker... Damn probably I was just lucky not to hit this issue, I did some test using `3.1.0-rev3` vs `3.1.0-rev4` vs `3.1.0` in the addons e2e tets and they all behaved as expected. I guess this happened because in those tests I did they all sort the same following alphabetical order.

I had the impression the versions were parsed as semantic versions somehow, I just verified that there was a library for such a thing in skuba...

Reply:
@ereslibre is the *version.Version object converted to a String once you use `>` ?

Reply:
Before filing the PR to fix version compare, I would like to discuss our image version (image tag + revision).

Currently, gangway bump version from `3.1.0-rev4` to `3.1.0` and I make some example tests at [here](https://github.com/jenting/semver/blob/master/example_test.go). It seems like the version comparison is not what I thought `3.1.0-rev4` is greater than `3.1.0`. Here is the reference to [semantic versioning comparision](https://semver.org/#spec-item-11)

I come to a workaround on fixing this issue is to compare the manifest version first, and if equals, compare image version.

Any ideas are welcome to come out. @ereslibre @innobead @MaximilianMeister @jordimassaguerpla @davidcassany

Reply:
Ouch! :disappointed: Yes I fact I remember I was afraid of this behavior and so that I run several tests last week of `skuba addon upgrade plan` with different combinations, however I did not realize that semantic versionig parser was not in use...

Reply:
How is the new addon thing working? Probably we could force an update of all images and use the revision number for all of them. Then if we always use the schema as `\d+(\.\d+){0,2}-rev\d+` we are safe. I am not a fan of this solution, but is the first that comes to my mind. Still thinking

Reply:
`semver` has some woes when it comes to this way of [comparing versions](https://semver.org/#spec-item-11), for our use case. Quoting:

> When major, minor, and patch are equal, a pre-release version has lower precedence than a normal version.

> Numeric identifiers always have lower precedence than non-numeric identifiers.

So I propose to do the following, what should always work:

1. Compare `manifestVersion`; if `new.manifestVersion <= old.manifestVersion`, we are done. Either the addon wasn't updated, or we are calling this from an older skuba.

2. If `old.manifestVersion == nil` and `new.manifestVersion != nil`, then it's a new addon.

3. If `new.manifestVersion > old.manifestVersion`, then this addon has an upgrade.

3.1 Compare (string comparison) of versions, no need to parse anything: If `new.AddonVersion != old.AddonVersion` show on the `skuba addon upgrade plan` output the old and new versions, like `%s => %s` with old.AddonVersion and new.AddonVeresion.

3.2 If `new.AddonVersion == old.AddonVersion`, then just show something like `%s (manifest version from %s to %s)` with `new.AddonVersion`, `old.manifestVersion`, `new.manifestVersion`.

So we always depend on the comparison of the int `manifestVersion` to determine whether we have a new version of the addon overall or not, and then just equality string comparisons to show old and new (since we are comparing with our `manifestVersion` we don't have to parse semvers at all, a version on an older manifestVersion **has** to be smaller).

Reply:
Let me open a PR.